src/expressions/method-call-expr.md

```diff
@@ -15,22 +15,155 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
+more types will be considered for the former usage than the latter usage.
+
+For instance, if the receiver has type `Box<[i32;2]>`, then the candidate types
+will be `Box<[i32;2]>`,`[i32; 2]` (by dereferencing), and `[i32]` (by unsized
+coercion).
+
+If `SmartPtr<T>: Receiver<Target=T>`, and the receiver type is `&SmartPtr<Foo>`,
+then the candidate types would be `&SmartPtr<Foo>`, `SmartPtr<Foo>` and `Foo`.
+
+## Determining candidate methods
+
+This list of candidate types is then converted to a list of candidate methods.
+For each step, the `self` type is used to determine what searches to perform:
+
+* For a trait object, there is first a search for for inherent candidates for
+  the object, then inherent impl candidates for the type.
+* For a struct, enum, or foreign type, there is a search for inherent
+  impl candidates for the type.
+* For a type param, there's a search for for inherent candidates on the param.
+* For other tyings (e.g. bools, chars) there's a search for inherent candidates
+  for the incoherent type.
+* After any of these, there's a further search for extension candidates for
+  traits in scope.
+
+These searches contribute to list of all the candidate methods found;
+separate lists are maintained for inherent and extension candidates. Only
+[visible] candidates are included.
+
+(For diagnostic purposes, the search may be performed slightly differently, for
+instance searching all traits not just those in scope, or also noting
+inaccessible candidates.)
+
+## Picking a method from the candidates
+
+Once the list of candidate methods is assembled, the "picking" process
+starts.
+
+Once again, the candidate types are iterated. This time, only those types
+are iterated which can be reached via the `Deref` trait or built-in derefs;
+as noted above, this may be a shorter list than those that can be reached
+using the `Receiver` trait.
+
+For each step, picking is attempted in this order:
+
+* First, a by-value method, where the `self` type precisely matches
+  * First for inherent methods
+  * Then for extension methods
+* Then, a method where `self` is received by immutable reference (`&T`)
+  * First for inherent methods
+  * Then for extension methods
+* Then, a method where `self` is received by mutable reference (`&mut T`)
+  * First for inherent methods
+  * Then for extension methods
+* Then, a method where the `self` type is a `*const T` - this is only considered
+  if the self type is `*mut T`)
+  * First for inherent methods
+  * Then for extension methods
+* And finally, a method with a `Pin` that's reborrowed, if the `pin_ergonomics`
+  feature is enabled.
+  * First for inherent methods
+  * Then for extension methods
+
+For each of those searches, if exactly one candidate is identified,
+it's picked, and the search stops. If this results in multiple possible candidates,
+then it is an error, and the receiver must be [converted][disambiguate call]
+to an appropriate receiver type to make the method call.
```

Comment:
```suggestion
then it is an error, and the user must [disambiguate][disambiguate call]
the call.
```

This is about clarifying that it's the user and not the compiler that must do something here.

Regarding the bit about conversion, it seems better to leave that off as it's not really about converting the receiver (i.e. the expression), it's about ascribing the receiver type in the syntax.